### ISSUE TITLE
Allow pre-rendering to be opt-in instead of opt-out per page

### ISSUE THREAD

**Issue by AaronBeaudoin** [brillout/vite-plugin-ssr#373]
### Description

Assume the following VPS config in `vite.config.js`:

```js
{
  prerender: {
    partial: true
  }
}
```

Currently, VPS will try and pre-render all pages that do not have `doNotPrerender = true` exported. I propose the `doNotPrerender` export is removed in favor of simply using the `prerender` export, which would be simpler in my humble opinion. Here's how it would work:

- VPS has a new config option `vite.config.js#plugins[].ssr().prerender.default` which defaults to `true`. This means that the default value for the `prerender` export for all pages is `true`, which in other words means that VPS will by default try to pre-render all pages.
- To "opt-out" of pre-rendering for a specific page, simply export `prerender = false`.
- To "opt-out" of pre-rendering by default, set the `prerender.default` config option to `false` instead. This means that the default value for the `prerender` export will now be `false`, which in other words means that VPS will by default not try and pre-render any pages.
- Now, to "opt-in" again to pre-rendering for a specific page, simply export `prerender = true`, or a function in the case of parameterized routes.

**Comment by brillout** [comment]
Good idea 👍 and should be possible. (There is only a slight design restriction but I think we can ignore it for now). I'll have a closer look at this tomorrow.

**Comment by brillout** [comment]
Actually, it turns out to require some brain power to implement this. (It's not just a renaming, as it clashes with the [`prerender()` hook](https://vite-plugin-ssr.com/prerender).)

I'm still open to implement this but I'm decreasing its priority. Closing in the meantime.

I believe that `prerender.partial` will cover the vast majority of use cases. If not, let me know in this ticket and I'll reconsider the priority of this ticket.

**Comment by AaronBeaudoin** [comment]
The only use cases that has a pain point is that in which the user has many pages and wishes to pre-render only a few of them. In such a case, they add `prerender: { partial: true }` to enable it globally, and then add `doNotPrerender` to **all** pages, then remove it for those they **do** wish to pre-render. It's "supported", but it's not really ergonomic.

Take ecommerce, for example. In most cases (in other words, _by default_) I wouldn't want to pre-render because generally my pages need to be dynamic, but in a few places I do, such as for marketing landing pages, or maybe strictly information pages, or maybe blog posts. In this case the best solution I have is to do the above: add `doNotPrerender` to _all_ my pages, then remove it from _some_ of them. Perhaps it's not a bad as it sounds, but it's still not really intuitive in my opinion.

So really the root of this idea is to somehow allow users to enable pre-rendering, but have it off by default. Then, they can export some variable that would _explicitly enable_ pre-rendering.

Low priority is fine since as you said this use cases is still possible; I believe this would be a nice improvement to increase the flexibility of VPS in the future.

**Comment by brillout** [comment]
Actually, one thing we could do is this:

```js
// /renderer/_default.page.server.js (or /_default/_default.page.server.js ;-))

export const doNotPrerender = true
```

This would essentially make pre-rendering opt-in. It's not supported, but I'll have a look at implementing this tomorrow.

Neat thing here:

```js
// /marketing/_default.page.server.js

// Enable pre-rendering for all marketing pages
export const doNotPrerender = false
```

See https://vite-plugin-ssr.com/file-structure#domain-driven.


So only two `doNotPrerender` usage are enough to achieve your use case.

Re-opening.

**Comment by AaronBeaudoin** [comment]
That is a **fantastic** idea.
